crates/storage/src/map.rs

```diff
@@ -425,3 +423,1304 @@ mod test {
+
+        // set and remove it
+        PEOPLE.save(&mut store, b"removed", &data).unwrap();
+        PEOPLE.remove(&mut store, b"removed");
+
+        // invalid, but non-empty data
+        store.write(&PEOPLE.path(b"random").storage_key, b"random-data");
+
+        // any data, including invalid or empty is returned as "has"
+        assert!(PEOPLE.has(&store, b"john"));
+        assert!(PEOPLE.has(&store, b"random"));
+
+        // if nothing was written, it is false
+        assert!(!PEOPLE.has(&store, b"never-writen"));
+        assert!(!PEOPLE.has(&store, b"removed"));
+    }
+
+    #[test]
+    fn composite_keys() {
+        let mut store = MockStorage::new();
+
+        // save and load on a composite key
+        let allow = ALLOWANCE.path((b"owner", b"spender"));
+        let allow = allow.as_path();
+        assert_eq!(None, allow.may_load(&store).unwrap());
+        allow.save(&mut store, &1234).unwrap();
+        assert_eq!(1234, allow.load(&store).unwrap());
+
+        // not under other key
+        let different = ALLOWANCE.may_load(&store, (b"owners", b"pender")).unwrap();
+        assert_eq!(None, different);
+
+        // matches under a proper copy
+        let same = ALLOWANCE.load(&store, (b"owner", b"spender")).unwrap();
+        assert_eq!(1234, same);
+    }
+
+    #[test]
+    fn triple_keys() {
+        let mut store = MockStorage::new();
+
+        // save and load on a triple composite key
+        let triple = TRIPLE.path((b"owner", 10u8, "recipient"));
+        let triple = triple.as_path();
+        assert_eq!(None, triple.may_load(&store).unwrap());
+        triple.save(&mut store, &1234).unwrap();
+        assert_eq!(1234, triple.load(&store).unwrap());
+
+        // not under other key
+        let different = TRIPLE
+            .may_load(&store, (b"owners", 10u8, "ecipient"))
+            .unwrap();
+        assert_eq!(None, different);
+
+        // matches under a proper copy
+        let same = TRIPLE.load(&store, (b"owner", 10u8, "recipient")).unwrap();
+        assert_eq!(1234, same);
+    }
+
+    #[test]
+    fn range_raw_simple_key() {
+        let mut store = MockStorage::new();
+
+        // save and load on two keys
+        let data1 = Data {
+            name: "John".to_string(),
+            age: 32,
+        };
+        PEOPLE.save(&mut store, b"john", &data1).unwrap();
+
+        let data2 = Data {
+            name: "Jim".to_string(),
+            age: 44,
+        };
+        PEOPLE.save(&mut store, b"jim", &data2).unwrap();
+
+        let data_1_raw = data1.to_borsh_vec().unwrap();
+        let data_2_raw = data2.to_borsh_vec().unwrap();
+
+        // let's try to iterate!
+        let all: Vec<_> = PEOPLE
+            .range_raw(&store, None, None, Order::Ascending)
+            .collect();
+
+        assert_eq!(2, all.len());
+        assert_eq!(all, vec![
+            (b"jim".to_vec(), data_2_raw.clone()),
+            (b"john".to_vec(), data_1_raw.clone())
+        ]);
```

Comment:
this can just be an array `[T; N]` instead of a `Vec<T>`, because `Vec<T>` implements `Partial<[T; N]>`.

Array is faster because it doesn't involve dynamic allocation.